File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex05AddressValidation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-05 — Validating the returned address.
/// </summary>
/// <remarks>
/// Validate the returned address structurally — present, self-consistent, or neither.
/// Complete is not the same as correct.
/// </remarks>
public sealed class Ex05AddressValidation : IExample
{
    /// <inheritdoc />
    public string Id => "ex-05";

    /// <inheritdoc />
    public string Title => "Check the returned address for presence and self-consistency.";

    /// <summary>Placeholders that are present, non-null, and mean nothing.</summary>
    private static readonly string[] Placeholders = { "n/a", "na", "none", "unknown", "-", "." };

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        // The inconsistent-address fixture is the point of this example, and production
        // will not produce one on request.
        using var context = ExampleContext.WithFixtures();

        foreach (var ein in new[] { Fixtures.Eins.PublicCharity, Fixtures.Eins.InconsistentAddress })
        {
            var result = await context.Client.Nonprofits.CheckAsync(ein);
            var organization = result.Nonprofit;

            if (organization is null)
            {
                continue;
            }

            Output.Heading(Output.Text(organization.OrganizationName));

            Output.DisplayField(organization, "address_line1");
            Output.DisplayField(organization, "address_line2");
            Output.DisplayField(organization, "city");
            Output.DisplayField(organization, "state");
            Output.DisplayField(organization, "state_name");
            Output.DisplayField(organization, "zip");

            var findings = new List<string>();

            // 1. Presence. A field the API did not return and a field it returned as null
            //    are different facts; neither is an address.
            foreach (var field in new[] { "address_line1", "city", "state", "zip" })
            {
                if (!organization.Has(field))
                {
                    findings.Add($"{field} was not returned");
                }
                else if (string.IsNullOrWhiteSpace(organization.GetString(field)))
                {
                    findings.Add($"{field} is null or empty");
                }
            }

            // 2. Placeholders. Present, non-null, and no more an address than a blank.
            foreach (var field in new[] { "address_line1", "address_line2", "city" })
            {
                var value = organization.GetString(field)?.Trim().ToLowerInvariant();

                if (value is not null && Placeholders.Contains(value, StringComparer.Ordinal))
                {
                    findings.Add($"{field} holds a placeholder");
                }
            }

            // 3. Self-consistency. The state code, the spelled-out state name and the
            //    ZIP prefix all describe the same thing, so they can be held against
            //    each other without any external data.
            var state = organization.State;
            var stateName = organization.StateName;

            if (state is not null && stateName is not null && !StateAgrees(state, stateName))
            {
                findings.Add($"state \"{state}\" and state_name \"{stateName}\" disagree");
            }

            // The Pub 78 copy of the city is a second opinion, free of charge.
            var cities = new[] { organization.City, organization.Pub78City }
                .Where(city => !string.IsNullOrWhiteSpace(city))
                .Select(city => city!.ToUpperInvariant())
                .Distinct(StringComparer.Ordinal)
                .ToList();

            if (cities.Count > 1)
            {
                findings.Add("the sources disagree on the city: " + string.Join(" / ", cities));
            }

            Output.Field("findings", findings.Count == 0 ? "none" : string.Join("; ", findings));
            Output.Field("routes to", findings.Count == 0 ? "accept" : "manual review");
        }

        Output.Note(
            "Structural checks only. That an address is present, non-placeholder and\n"
            + "self-consistent says nothing about whether anyone is there — for that you\n"
            + "need an address verification service, which this API is not.");
    }

    /// <summary>
    /// Whether a two-letter code and a spelled-out name describe the same state.
    /// </summary>
    /// <remarks>
    /// A handful of pairs, enough to make the point. A production check would use a
    /// complete table; the shape of the check is what matters here.
    /// </remarks>
    private static bool StateAgrees(string code, string name)
    {
        var known = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["MA"] = "Massachusetts",
            ["ME"] = "Maine",
            ["NY"] = "New York",
            ["CA"] = "California",
            ["TX"] = "Texas",
        };

        // An unknown code is not a disagreement — it is a table this example does not have.
        return !known.TryGetValue(code, out var expected)
            || string.Equals(expected, name, StringComparison.OrdinalIgnoreCase);
    }
}

File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex05AddressValidation.java
package org.pactman.nonprofitcheckplus.examples.cases;

import java.util.List;
import org.pactman.nonprofitcheckplus.devtools.FixtureEins;
import org.pactman.nonprofitcheckplus.examples.support.Addresses;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.models.Nonprofit;

/**
 * EX-05 — Address validation.
 *
 * <p>A non-null address field is not a usable address. This example runs three
 * records through the same checks: one clean, one with several fields returned
 * as null, and one where every component is present and they contradict each
 * other.
 */
public final class Ex05AddressValidation implements Example {

    @Override
    public String id() {
        return "ex-05";
    }

    @Override
    public String title() {
        return "Address validation beyond a null check";
    }

    @Override
    public void run(String[] args) {
        try (ExampleContext context = ExampleContext.withFixtures()) {
            for (String ein : new String[] {
                FixtureEins.PUBLIC_CHARITY,
                FixtureEins.SPARSE_IDENTITY,
                FixtureEins.INCONSISTENT_ADDRESS,
            }) {
                Nonprofit nonprofit = context.client().nonprofits().check(ein).getNonprofit();

                Output.heading(nonprofit.getOrganizationName() + " (" + ein + ")");
                Output.field("address_line1", Output.present(nonprofit, "address_line1"));
                Output.field("address_line2", Output.present(nonprofit, "address_line2"));
                Output.field("city", Output.present(nonprofit, "city"));
                Output.field("state", Output.present(nonprofit, "state"));
                Output.field("state_name", Output.present(nonprofit, "state_name"));
                Output.field("zip", Output.present(nonprofit, "zip"));

                List<String> findings = Addresses.findings(nonprofit);

                System.out.println();

                if (findings.isEmpty()) {
                    Output.bullet("nothing to report");
                } else {
                    for (String finding : findings) {
                        Output.bullet(finding);
                    }
                }
            }
        }

        Output.note("The third record would pass any check that only asks whether the fields\n"
                + "came back non-null: every component is there. The state code, the spelled-out\n"
                + "state and the ZIP describe three different places, and address_line2 holds a\n"
                + "placeholder. Read the values, not just their presence.");
    }
}
